### ISSUE TITLE
Understand top-p vs top-k

### ISSUE THREAD

**Issue by astariul** [rusiaaman/XLnet-gen#2]
I'm trying to understand top-p and top-k prediction mode from your code.
(I'm not proficient in Tensorflow and I never heard of top-p before)

---

Top-k strategy will sample token from the K best logits score.

Top-p strategy will sample token from the X logits where `logits score > p`.

**Am I right ?**

**Comment by rusiaaman** [comment]
top-p strategy here is the nucleus sampling first introduced in https://arxiv.org/abs/1904.09751. This has shown to increase the diversity in language generation. It is not `logits score>p` it is like sampling from a multi-nomial distribution where only those top probable tokens are considered which cumulatively have probability p.
